<div align="right">

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/wisaaco/TallerPythonFEE/blob/main/lessons/1/2_PandasIntroduccio.ipynb)

</div>

# 2 - La llibreria Pandas

## Pandas

**Pandas** és la llibreria per excel·lència per a l'anàlisi de dades del llenguatge Python. El seu nom prové de “panell data” (terme economètric). Inspirada en les funcionalitats de fulls de càlcul i R, però amb el potencial d'aquest llenguatge de propòsit general.

**Pandas** inclou totes les funcionalitats necessàries per al procés d'anàlisi de dades: càrrega, filtrat, tractament, síntesi, agrupament, emmagatzematge i visualització. A més, s'integra amb la resta de llibreries de càlcul numèric com Numpy, visualització amb Matplotlib, aprenentatge automàtic amb scikit-learn, etc.

**L'objectiu d'aquest curs introductori és adquirir capacitats bàsiques d'anàlisis de dades amb Pandas.**

Referències:
- https://pandas.pydata.org/
- https://en.wikipedia.org/wiki/Python_(programming_language)
- https://github.com/pandas-dev/pandas#*main-*features

Nota:
- Pandas és una llibrería instal·lada a Google Colab.

In [1]:
import pandas as pd

## Càrrega de dades i estructures bàsiques de Pandas.


A Pandas hi ha dos tipus de variables: dataframes i series.

In [2]:
df_who = pd.read_csv("data/WHO.csv") #dataframe
print(type(df_who))

<class 'pandas.core.frame.DataFrame'>


In [3]:
df_who

Country  CountryID  Continent  Adolescent fertility rate (%)  \
0           Afghanistan          1          1                          151.0   
1               Albania          2          2                           27.0   
2               Algeria          3          3                            6.0   
3               Andorra          4          2                            NaN   
4                Angola          5          3                          146.0   
..                  ...        ...        ...                            ...   
197             Vietnam        198          6                           25.0   
198  West Bank and Gaza        199          1                            NaN   
199               Yemen        200          1                           83.0   
200              Zambia        201          3                          161.0   
201            Zimbabwe        202          3                          101.0   

     Adult literacy rate (%)  \
0                       28.0   
1                       98.7   
2                       69.9   
3                        NaN   
4                       67.4   
..                       ...   
197                     90.3   
198                      NaN   
199                     54.1   
200                     68.0   
201                     89.5   

     Gross national income per capita (PPP international $)  \
0                                                  NaN        
1                                               6000.0        
2                                               5940.0        
3                                                  NaN        
4                                               3890.0        
..                                                 ...        
197                                             2310.0        
198                                                NaN        
199                                             2090.0        
200                                             1140.0        
201                                                NaN        

     Net primary school enrolment ratio female (%)  \
0                                              NaN   
1                                             93.0   
2                                             94.0   
3                                             83.0   
4                                             49.0   
..                                             ...   
197                                           91.0   
198                                            NaN   
199                                           65.0   
200                                           94.0   
201                                           88.0   

     Net primary school enrolment ratio male (%)  \
0                                            NaN   
1                                           94.0   
2                                           96.0   
3                                           83.0   
4                                           51.0   
..                                           ...   
197                                         96.0   
198                                          NaN   
199                                         85.0   
200                                         90.0   
201                                         87.0   

     Population (in thousands) total  Population annual growth rate (%)  ...  \
0                            26088.0                                4.0  ...   
1                             3172.0                                0.6  ...   
2                            33351.0                                1.5  ...   
3                               74.0                                1.0  ...   
4                            16557.0                                2.8  ...   
..                               ...                                ...  ...   
197                          86206.0                                1.4  ...   

In [4]:
# Estructura
print(df_who.shape) # files o mostres x columnes

(202, 358)


In [5]:
df_who.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202 entries, 0 to 201
Columns: 358 entries, Country to Urban_population_pct_of_total
dtypes: float64(355), int64(2), object(1)
memory usage: 565.1+ KB


In [6]:
df_who.describe()

CountryID   Continent  Adolescent fertility rate (%)  \
count  202.000000  202.000000                     177.000000   
mean   101.500000    3.579208                      59.457627   
std     58.456537    1.808263                      49.105286   
min      1.000000    1.000000                       0.000000   
25%     51.250000    2.000000                      19.000000   
50%    101.500000    3.000000                      46.000000   
75%    151.750000    5.000000                      91.000000   
max    202.000000    7.000000                     199.000000   

       Adult literacy rate (%)  \
count               131.000000   
mean                 78.871756   
std                  20.415760   
min                  23.600000   
25%                  68.400000   
50%                  86.500000   
75%                  95.300000   
max                  99.800000   

       Gross national income per capita (PPP international $)  \
count                                         178.000000        
mean                                        11250.112360        
std                                         12586.753417        
min                                           260.000000        
25%                                          2112.500000        
50%                                          6175.000000        
75%                                         14502.500000        
max                                         60870.000000        

       Net primary school enrolment ratio female (%)  \
count                                     179.000000   
mean                                       84.033520   
std                                        17.788047   
min                                         6.000000   
25%                                        79.000000   
50%                                        90.000000   
75%                                        96.000000   
max                                       100.000000   

       Net primary school enrolment ratio male (%)  \
count                                   179.000000   
mean                                     85.698324   
std                                      15.451212   
min                                      11.000000   
25%                                      79.500000   
50%                                      90.000000   
75%                                      96.000000   
max                                     100.000000   

       Population (in thousands) total  Population annual growth rate (%)  \
count                     1.930000e+02                         193.000000   
mean                      3.409805e+04                           1.297927   
std                       1.304957e+05                           1.163864   
min                       2.000000e+00                          -2.500000   
25%                       1.340000e+03                           0.500000   
50%                       6.762000e+03                           1.300000   
75%                       2.173200e+04                           2.100000   
max                       1.328474e+06                           4.300000   

       Population in urban areas (%)  ...  Total_CO2_emissions  Total_income  \
count                     193.000000  ...         1.860000e+02  1.780000e+02   
mean                       54.911917  ...         1.483596e+05  2.015567e+11   
std                        23.554182  ...         6.133091e+05  9.400689e+11   
min                        10.000000  ...         2.565000e+01  5.190000e+07   
25%                        36.000000  ...         1.672615e+03  3.317500e+09   
50%                        57.000000  ...         1.021157e+04  1.145000e+10   
75%                        73.000000  ...         6.549217e+04  8.680000e+10   
max                       100.000000  ...         5.776432e+06  1.100000e+13   

       Total_reserves  Trade_balance_goods_and_services  \
count      128.000000                      1.710000e+02   
mean        5

In [7]:
df_who.head()

Country  CountryID  Continent  Adolescent fertility rate (%)  \
0  Afghanistan          1          1                          151.0   
1      Albania          2          2                           27.0   
2      Algeria          3          3                            6.0   
3      Andorra          4          2                            NaN   
4       Angola          5          3                          146.0   

   Adult literacy rate (%)  \
0                     28.0   
1                     98.7   
2                     69.9   
3                      NaN   
4                     67.4   

   Gross national income per capita (PPP international $)  \
0                                                NaN        
1                                             6000.0        
2                                             5940.0        
3                                                NaN        
4                                             3890.0        

   Net primary school enrolment ratio female (%)  \
0                                            NaN   
1                                           93.0   
2                                           94.0   
3                                           83.0   
4                                           49.0   

   Net primary school enrolment ratio male (%)  \
0                                          NaN   
1                                         94.0   
2                                         96.0   
3                                         83.0   
4                                         51.0   

   Population (in thousands) total  Population annual growth rate (%)  ...  \
0                          26088.0                                4.0  ...   
1                           3172.0                                0.6  ...   
2                          33351.0                                1.5  ...   
3                             74.0                                1.0  ...   
4                          16557.0                                2.8  ...   

   Total_CO2_emissions  Total_income  Total_reserves  \
0               692.50           NaN             NaN   
1              3499.12  4.790000e+09           78.14   
2            137535.56  6.970000e+10          351.36   
3                  NaN           NaN             NaN   
4              8991.46  1.490000e+10           27.13   

   Trade_balance_goods_and_services  Under_five_mortality_from_CME  \
0                               NaN                         257.00   
1                     -2.040000e+09                          18.47   
2                      4.700000e+09                          40.00   
3                               NaN                            NaN   
4                      9.140000e+09                         164.10   

   Under_five_mortality_from_IHME  Under_five_mortality_rate  \
0                           231.9                     257.00   
1                            15.5                      18.47   
2                            31.2                      40.00   
3                             NaN                        NaN   
4                           242.5                     164.10   

   Urban_population  Urban_population_growth  Urban_population_pct_of_total  
0         5740436.0                     5.44                           22.9  
1         1431793.9                     2.21                           45.4  
2        20800000.0                     2.61                           63.3  
3               NaN                      NaN                            NaN  
4         8578749.0                     4.14                           53.3  

[5 rows x 358 columns]

In [8]:
df_who.tail()

Country  CountryID  Continent  Adolescent fertility rate (%)  \
197             Vietnam        198          6                           25.0   
198  West Bank and Gaza        199          1                            NaN   
199               Yemen        200          1                           83.0   
200              Zambia        201          3                          161.0   
201            Zimbabwe        202          3                          101.0   

     Adult literacy rate (%)  \
197                     90.3   
198                      NaN   
199                     54.1   
200                     68.0   
201                     89.5   

     Gross national income per capita (PPP international $)  \
197                                             2310.0        
198                                                NaN        
199                                             2090.0        
200                                             1140.0        
201                                                NaN        

     Net primary school enrolment ratio female (%)  \
197                                           91.0   
198                                            NaN   
199                                           65.0   
200                                           94.0   
201                                           88.0   

     Net primary school enrolment ratio male (%)  \
197                                         96.0   
198                                          NaN   
199                                         85.0   
200                                         90.0   
201                                         87.0   

     Population (in thousands) total  Population annual growth rate (%)  ...  \
197                          86206.0                                1.4  ...   
198                              NaN                                NaN  ...   
199                          21732.0                                3.0  ...   
200                          11696.0                                1.9  ...   
201                          13228.0                                0.8  ...   

     Total_CO2_emissions  Total_income  Total_reserves  \
197            101826.23  4.480000e+10           47.11   
198               655.86  3.780000e+09             NaN   
199             20148.34  1.150000e+10          114.52   
200              2366.94  4.090000e+09           10.41   
201             11457.33  5.620000e+09            3.39   

     Trade_balance_goods_and_services  Under_five_mortality_from_CME  \
197                     -1.940000e+09                           20.2   
198                               NaN                           28.0   
199                      8.310000e+08                           82.4   
200                     -4.470000e+08                          175.3   
201                     -1.710000e+08                          106.5   

     Under_five_mortality_from_IHME  Under_five_mortality_rate  \
197                            23.4                       20.2   
198                            25.8                       28.0   
199                            87.9                       82.4   
200                           163.8                      175.3   
201                            67.0                      106.5   

     Urban_population  Urban_population_growth  Urban_population_pct_of_total  
197        21900000.0                     2.90                           26.4  
198         2596216.0                     3.33                           71.6  
199         5759120.5                     4.37                           27.3  
200         4017411.0                     1.95                           35.0  
201         4709965.0                     1.90                           35.9  

[5 rows x 358 columns]

### Columnes

In [9]:
# Columnes o caracteristiques de cada mostra
print(df_who.columns)

Index(['Country', 'CountryID', 'Continent', 'Adolescent fertility rate (%)',
       'Adult literacy rate (%)',
       'Gross national income per capita (PPP international $)',
       'Net primary school enrolment ratio female (%)',
       'Net primary school enrolment ratio male (%)',
       'Population (in thousands) total', 'Population annual growth rate (%)',
       ...
       'Total_CO2_emissions', 'Total_income', 'Total_reserves',
       'Trade_balance_goods_and_services', 'Under_five_mortality_from_CME',
       'Under_five_mortality_from_IHME', 'Under_five_mortality_rate',
       'Urban_population', 'Urban_population_growth',
       'Urban_population_pct_of_total'],
      dtype='object', length=358)


In [10]:
for col in df_who.columns:
    print(col)

Country
CountryID
Continent
Adolescent fertility rate (%)
Adult literacy rate (%)
Gross national income per capita (PPP international $)
Net primary school enrolment ratio female (%)
Net primary school enrolment ratio male (%)
Population (in thousands) total
Population annual growth rate (%)
Population in urban areas (%)
Population living below the poverty line (% living on &lt; US$1 per day)
Population median age (years)
Population proportion over 60 (%)
Population proportion under 15 (%)
Registration coverage of births (%)
Total fertility rate (per woman)
Antenatal care coverage - at least four visits (%)
Antiretroviral therapy coverage among HIV-infected pregt women for PMTCT (%)
Antiretroviral therapy coverage among people with advanced HIV infections (%)
Births attended by skilled health personnel (%)
Births by caesarean section (%)
Children aged 6-59 months who received vitamin A supplementation (%)
Children aged &lt;5 years sleeping under insecticide-treated nets (%)
Children ag

In [11]:
df_who["Country"] # Una serie

0             Afghanistan
1                 Albania
2                 Algeria
3                 Andorra
4                  Angola
              ...        
197               Vietnam
198    West Bank and Gaza
199                 Yemen
200                Zambia
201              Zimbabwe
Name: Country, Length: 202, dtype: object

In [12]:
print(df_who["Country"][0])
print("-"*30)
print(df_who["Country"][:5])
print("-"*30)
print(df_who["Country"].values)

Afghanistan
------------------------------
0    Afghanistan
1        Albania
2        Algeria
3        Andorra
4         Angola
Name: Country, dtype: object
------------------------------
['Afghanistan' 'Albania' 'Algeria' 'Andorra' 'Angola'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Australia' 'Austria'
 'Azerbaijan' 'Bahamas' 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus'
 'Belgium' 'Belize' 'Benin' 'Bermuda' 'Bhutan' 'Bolivia'
 'Bosnia and Herzegovina' 'Botswana' 'Brazil' 'Brunei Darussalam'
 'Bulgaria' 'Burkina Faso' 'Burundi' 'Cambodia' 'Cameroon' 'Canada'
 'Cape Verde' 'Central African Republic' 'Chad' 'Chile' 'China' 'Colombia'
 'Comoros' 'Congo, Dem. Rep.' 'Congo, Rep.' 'Cook Islands' 'Costa Rica'
 "Cote d'Ivoire" 'Croatia' 'Cuba' 'Cyprus' 'Czech Republic' 'Denmark'
 'Djibouti' 'Dominica' 'Dominican Republic' 'Ecuador' 'Egypt'
 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Ethiopia' 'Fiji'
 'Finland' 'France' 'French Polynesia' 'Gabon' 'Gambia' 'Georgia'
 'Germany' 'Gh

In [13]:
df_who.Country # una altra manera més simple
# pero hi ha noms de columnes molt llargues: "Children aged &lt;5 years who received any antimalarial treatment for fever (%)" 
# NOTA: En la creació de documents és important un adequat nom de columnes

0             Afghanistan
1                 Albania
2                 Algeria
3                 Andorra
4                  Angola
              ...        
197               Vietnam
198    West Bank and Gaza
199                 Yemen
200                Zambia
201              Zimbabwe
Name: Country, Length: 202, dtype: object

In [14]:
print(df_who.columns[9])
print("-"*30)
print(df_who[df_who.columns[9]])


Population annual growth rate (%)
------------------------------
0      4.0
1      0.6
2      1.5
3      1.0
4      2.8
      ... 
197    1.4
198    NaN
199    3.0
200    1.9
201    0.8
Name: Population annual growth rate (%), Length: 202, dtype: float64


In [15]:
# Es podem agagar multiples columnes
df_who[df_who.columns[0:5]]


Country  CountryID  Continent  Adolescent fertility rate (%)  \
0           Afghanistan          1          1                          151.0   
1               Albania          2          2                           27.0   
2               Algeria          3          3                            6.0   
3               Andorra          4          2                            NaN   
4                Angola          5          3                          146.0   
..                  ...        ...        ...                            ...   
197             Vietnam        198          6                           25.0   
198  West Bank and Gaza        199          1                            NaN   
199               Yemen        200          1                           83.0   
200              Zambia        201          3                          161.0   
201            Zimbabwe        202          3                          101.0   

     Adult literacy rate (%)  
0                       28.0  
1                       98.7  
2                       69.9  
3                        NaN  
4                       67.4  
..                       ...  
197                     90.3  
198                      NaN  
199                     54.1  
200                     68.0  
201                     89.5  

[202 rows x 5 columns]

In [16]:
df_who[["CountryID","Continent"]]

CountryID  Continent
0            1          1
1            2          2
2            3          3
3            4          2
4            5          3
..         ...        ...
197        198          6
198        199          1
199        200          1
200        201          3
201        202          3

[202 rows x 2 columns]

In [17]:
# Podem accedir a les columnes amb la seva posició en lloc del nom
# funció: .iloc(files, columnes)
df_who.iloc[:,3:5]

Adolescent fertility rate (%)  Adult literacy rate (%)
0                            151.0                     28.0
1                             27.0                     98.7
2                              6.0                     69.9
3                              NaN                      NaN
4                            146.0                     67.4
..                             ...                      ...
197                           25.0                     90.3
198                            NaN                      NaN
199                           83.0                     54.1
200                          161.0                     68.0
201                          101.0                     89.5

[202 rows x 2 columns]

In [18]:
df_who.iloc[:,[3,5,9]]

Adolescent fertility rate (%)  \
0                            151.0   
1                             27.0   
2                              6.0   
3                              NaN   
4                            146.0   
..                             ...   
197                           25.0   
198                            NaN   
199                           83.0   
200                          161.0   
201                          101.0   

     Gross national income per capita (PPP international $)  \
0                                                  NaN        
1                                               6000.0        
2                                               5940.0        
3                                                  NaN        
4                                               3890.0        
..                                                 ...        
197                                             2310.0        
198                                                NaN        
199                                             2090.0        
200                                             1140.0        
201                                                NaN        

     Population annual growth rate (%)  
0                                  4.0  
1                                  0.6  
2                                  1.5  
3                                  1.0  
4                                  2.8  
..                                 ...  
197                                1.4  
198                                NaN  
199                                3.0  
200                                1.9  
201                                0.8  

[202 rows x 3 columns]

### Files

Cada mostra té un índex. L'índex pot ser numèric, alfabètic o de temps.

In [19]:
df_who.index

RangeIndex(start=0, stop=202, step=1)

In [20]:
df_who.index.values

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

In [21]:
df_who.loc[0] #la mostra zero

Country                           Afghanistan
CountryID                                   1
Continent                                   1
Adolescent fertility rate (%)           151.0
Adult literacy rate (%)                  28.0
                                     ...     
Under_five_mortality_from_IHME          231.9
Under_five_mortality_rate               257.0
Urban_population                    5740436.0
Urban_population_growth                  5.44
Urban_population_pct_of_total            22.9
Name: 0, Length: 358, dtype: object

In [22]:
type(df_who.loc[0]) # una mostra i una columna són també sèries

pandas.core.series.Series

In [23]:
print(df_who.loc[0].Country) # Podem accedir a una sèrie amb el seu índex (pròximes classes)
print(df_who.loc[0][0]) #o amb la seva posició
print(df_who.loc[0][3])

Afghanistan
Afghanistan
151.0


In [24]:
df_who.loc[166:170] #slicing

Country  CountryID  Continent  Adolescent fertility rate (%)  \
166       Somalia        167          3                           35.0   
167  South Africa        168          3                           65.0   
168         Spain        169          2                           10.0   
169     Sri Lanka        170          7                           29.0   
170         Sudan        171          3                           51.0   

     Adult literacy rate (%)  \
166                      NaN   
167                     82.4   
168                     97.2   
169                     90.7   
170                     60.9   

     Gross national income per capita (PPP international $)  \
166                                                NaN        
167                                             8900.0        
168                                            28200.0        
169                                             3730.0        
170                                             1780.0        

     Net primary school enrolment ratio female (%)  \
166                                            6.0   
167                                           88.0   
168                                           99.0   
169                                          100.0   
170                                           37.0   

     Net primary school enrolment ratio male (%)  \
166                                         11.0   
167                                         88.0   
168                                        100.0   
169                                         99.0   
170                                         45.0   

     Population (in thousands) total  Population annual growth rate (%)  ...  \
166                           8445.0                                3.0  ...   
167                          48282.0                                0.7  ...   
168                          43887.0                                1.1  ...   
169                          19207.0                                0.5  ...   
170                          37707.0                                2.2  ...   

     Total_CO2_emissions  Total_income  Total_reserves  \
166               589.90           NaN             NaN   
167            408792.47  1.610000e+11           66.32   
168            343701.53  6.780000e+11             NaN   
169             11017.65  1.980000e+10           24.28   
170             10618.27  1.670000e+10           10.13   

     Trade_balance_goods_and_services  Under_five_mortality_from_CME  \
166                               NaN                          150.8   
167                     -2.200000e+09                           68.5   
168                     -5.770000e+10                            4.9   
169                     -2.180000e+09                           21.4   
170                     -2.850000e+09                          111.3   

     Under_five_mortality_from_IHME  Under_five_mortality_rate  \
166                           219.9                      150.8   
167                            68.5                       68.5   
168                             4.2                        4.9   
169                            12.1                       21.4   
170                            94.4                      111.3   

     Urban_population  Urban_population_growth  Urban_population_pct_of_total  
166         2885131.0                     4.09                           35.2  
167        27800000.0                     1.98                           59.3  
168        33300000.0                     1.75                           76.7  
169         2969868.0                     0.26                           15.1  
170        15100000.0                     4.40                           40.8  

[5 rows x 358 columns]

In [25]:
# Els índexs de les mostres són útils quan són dates: Sèries temporals
# per exemple:
# df.loc["2020":"2022"]

In [26]:
# Obviament, podem fer una selecció més especifica
df_who.loc[167:169,["Country","Total_CO2_emissions"]] # per noms 

Country  Total_CO2_emissions
167  South Africa            408792.47
168         Spain            343701.53
169     Sri Lanka             11017.65

In [27]:
df_who.iloc[[168,192],[0,43,118]] # per posicions

Country  Hospital beds (per 10 000 population)  \
168                     Spain                                   34.0   
192  United States of America                                   32.0   

     Deaths among children under five years of age due to pneumonia (%)  
168                                                1.3                   
192                                                1.3

### Carrega de dades (II)

Per desconeixement encara s'utilitzin formats de codificació no oberts

In [28]:
# Data sources: https://datos.gob.es/es/catalogo?q=bilbao&g-recaptcha-response=&administration_level=L&theme_id=economia&sort=score+desc%2C+metadata_created+desc
# read_csv signature: https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

df_gastos = pd.read_csv("data/presupuesto_gastos_2023.csv")

ParserError: Error tokenizing data. C error: Expected 2 fields in line 8, saw 4


In [ ]:
df_gastos = pd.read_csv("data/presupuesto_gastos_2023.csv",delimiter=";")

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd3 in position 17: invalid continuation byte

### Codificació de caràcters

```python
var = "camión"
var = "lul·lià"
var = "Ζεύς"
var = "ประเทศไทย"
var = "日本語で"
```

Codificacions: 
- https://docs.python.org/3.11/library/codecs.html#standard-encodings
- UTF-8 https://es.wikipedia.org/wiki/UTF-8



In [ ]:
# cp1250 | windows-1250 | Central and Eastern Europe

# Mètode bàsic: "Prova i error" !!!
df_gastos = pd.read_csv("data/presupuesto_gastos_2023.csv",delimiter=";",encoding="cp1250")


In [ ]:
df_gastos.head()

_id SOZIETATEA_EU/SOCIEDAD_EU SOZIETATEA_CAS/SOCIEDAD_CAS  \
0    1                      UDAL                        UDAL   
1    2                      UDAL                        UDAL   
2    3                      UDAL                        UDAL   
3    4                      UDAL                        UDAL   
4    5                      UDAL                        UDAL   

   EKITALDIA/EJERCICIO SAILA/DEPARTAMENTO  \
0                 2023                110   
1                 2023                110   
2                 2023                110   
3                 2023                110   
4                 2023                110   

  SAILAREN DESKRIBAPENA_EU/DESCRIPCION DEPARTAMENTO_EU  \
0                            KULTURA ETA GOBERNANTZA     
1                            KULTURA ETA GOBERNANTZA     
2                            KULTURA ETA GOBERNANTZA     
3                            KULTURA ETA GOBERNANTZA     
4                            KULTURA ETA GOBERNANTZA     

  SAILAREN DESKRIBAPENA_EU/DESCRIPCION DEPARTAMENTO_CAS  \
0                               CULTURA Y GOBERNANZA      
1                               CULTURA Y GOBERNANZA      
2                               CULTURA Y GOBERNANZA      
3                               CULTURA Y GOBERNANZA      
4                               CULTURA Y GOBERNANZA      

  ZENTRU KUDEATZAILEA/CENTRO GESTOR  \
0                              1100   
1                              1100   
2                              1100   
3                              1100   
4                              1100   

  ZENTRO KUDEATZAILEAREN DESKR._EUS/DESCR. CENTRO GESTOR_EUS  \
0                            KULTURA ETA GOBERNANTZA           
1                            KULTURA ETA GOBERNANTZA           
2                            KULTURA ETA GOBERNANTZA           
3                            KULTURA ETA GOBERNANTZA           
4                            KULTURA ETA GOBERNANTZA           

  ZENTRO KUDEATZAILEAREN DESKR._CAS/DESCR. CENTRO GESTOR_CAS  ...  \
0                               CULTURA Y GOBERNANZA          ...   
1                               CULTURA Y GOBERNANZA          ...   
2                               CULTURA Y GOBERNANZA          ...   
3                               CULTURA Y GOBERNANZA          ...   
4                               CULTURA Y GOBERNANZA          ...   

   ARTIKULUAREN DESKRIBAPENA_CAS/DESCRIPCION ARTICULO_CAS KONTZEPTUA/CONCEPTO  \
0                      RETRIBUCIONES DE ALTOS CARGOS                      100   
1   RETRIBUCIONES DEL PERSONAL EVENTUAL DE GABINETES                      110   
2             RETRIBUCIONES DEL PERSONAL FUNCIONARIO                      120   
3             RETRIBUCIONES DEL PERSONAL FUNCIONARIO                      121   
4             RETRIBUCIONES DEL PERSONAL FUNCIONARIO                      121   

  KONTZEPTUAREN DESKRIBAPENA_EUS/DESCRIPCION CONCEPTO_EUS  \
0  GOI KARGUEN OINARRIZKO SOLDATAK ETA BESTELAKO ...        
1  KABINETEETAKO ALDI BAT. PERTSONALAREN OINARRIZ...        
2                 FUNTZIONARIOEN OINARRIZKO SOLDATAK        
3               FUNTZIONARIOEN ORDAINSARI OSAGARRIAK        
4               FUNTZIONARIOEN ORDAINSARI OSAGARRIAK        

   KONTZEPTUAREN DESKRIBAPENA_CAS/DESCRIPCION CONCEPTO_CAS  \
0  RETRIBUCIONES BASICAS Y OTRAS REMUNERACIONES D...         
1  RETRIBUCIONES BASICAS Y OTRAS REMUNERACIONES D...         
2     RETRIBUCIONES BASICAS DEL PERSONAL FUNCIONARIO         
3  RETRIBUCIONES COMPLEMENTARIAS DEL PERSONAL FUN...         
4  RETRIBUCIONES COMPLEMENTARIAS DEL PERSONAL FUN...         

  AZPIKONTZEPTUA/SUBCONCEPTO  \
0                      10001   
1                      11001   
2                      12001   
3                      12101   
4                      12102   

  AZPIKONTZEPTUAREN DESKRIBAPENA_EUS/DESCRIPCION SUBCONCEPTO_EUS  \
0  GOI KARGUEN OINARRIZKO SOLDATAK ETA BESTELAKO ...               
1  KABINETEETAKO ALDI BAT. PERTSONALAREN OINARRIZ...

### Guardant un dataframe

In [ ]:
df_gastos.to_csv('data/tmp_file.csv',encoding='utf-8') # guardant un dataframe en un fitxer, especificant el format


In [ ]:
df_tmp = pd.read_csv("data/tmp_file.csv") # test codificació i separador

### Carrega de dades (III)

Pandas pot cargar diferents formats de fitxers (csv,json, excel, etc.)
- https://pandas.pydata.org/docs/reference/io.html

In [51]:
url = "https://es.wikipedia.org/wiki/Anexo:Comunidades_y_ciudades_aut%C3%B3nomas_de_Espa%C3%B1a"
comunidades_esp = pd.io.html.read_html(url)
comunidades_esp # Alerta ! Encara no és un dataframe! En una web pots trobar més "taules"

[   Unnamed: 0                      Nombre  \
 0           1                   Andalucía   
 1           2                    Cataluña   
 2           3         Comunidad de Madrid   
 3           4        Comunidad Valenciana   
 4           5                     Galicia   
 5           6             Castilla y León   
 6           7                  País Vasco   
 7           8                    Canarias   
 8           9          Castilla-La Mancha   
 9          10            Región de Murcia   
 10         11                      Aragón   
 11         12              Islas Baleares   
 12         13                 Extremadura   
 13         14      Principado de Asturias   
 14         15  Comunidad Foral de Navarra   
 15         16                   Cantabria   
 16         17                    La Rioja   
 17         18                     Melilla   
 18         19                       Ceuta   
 19      TOTAL                      España   
 
            Capital (de iure o e

In [53]:
print(type(comunidades_esp[0]))
df_comunidades_esp = comunidades_esp[0]

<class 'pandas.core.frame.DataFrame'>


In [55]:
df_comunidades_esp.head() # funciona perfectament??

Unnamed: 0                Nombre Capital (de iure o en su defecto de facto)  \
0          1             Andalucía                                    Sevilla   
1          2              Cataluña                                  Barcelona   
2          3   Comunidad de Madrid                                     Madrid   
3          4  Comunidad Valenciana                                   Valencia   
4          5               Galicia                     Santiago de Compostela   

   Población (2022) Porcentaje población  Densidad (hab./km²)  \
0               NaN              17,87 %                 9705   
1               NaN              16,38 %                23884   
2               NaN              14,24 %                84115   
3               NaN              10,67 %                21698   
4               NaN               5,68 %                 9119   

  Superficie (km²) Porcentaje superficie Mapa  PIB per cápita en € (2021)  
0              NaN               17,31 %  NaN                         NaN  
1              NaN                6,35 %  NaN                         NaN  
2              NaN                1,59 %  NaN                         NaN  
3              NaN                4,60 %  NaN                         NaN  
4              NaN                5,84 %  NaN                         NaN

## Activitat

En aquesta activitat practicarem la càrrega de dades en diferents formats. Al món real les dades no sempre tenen una estructura i un format com ens agradaríem.

L'objectiu és que analitzeu la càrrega d'aquestes dades amb les originals: <br/>
- Quina dimensió tenen les dades reals i carregades?
- Quines columnes?
- El concepte de columna com atribut o característica i el concepte de fila com a mostra està present en l'estructura de les dades?
- Coincideixen amb la informació del fitxer?

### A) "Municipios y fenómeno demográfico de les Illes Balears" <br/>
Font: https://datos.gob.es/es/catalogo/ea0010587-balears-illes-por-municipios-y-fenomeno-demografico-mnpd-identificador-api-t20-e301-fenom-a2020-l0-23007-px <br/>
Fitxer: "data/municipis.csv" <br/>



In [86]:
import pandas as pd
#TODO

### B) "Speculation Watch List"
Font: https://data.cityofnewyork.us/Housing-Development/Speculation-Watch-List/adax-9mit <br/>
Fitxer: "data/Speculation_Watch_List.csv"

In [87]:
import pandas as pd
#TODO

### C) "Europe Inflation monthly data (annual rate of change)" 
Font: https://ec.europa.eu/eurostat/databrowser/view/PRC_HICP_MANR__custom_3761882/bookmark/table?lang=en&bookmarkId=4ad27e6f-358a-4a3d-82a0-587d69a833eb <br/>
Fitxer: "data/prc_hicp_manr__custom_3761882_spreadsheet.xlsx" <br/>
*Nota:*
  - Com podem agafar una "fulla" concreta del full de càlcul? (https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html)
  - Per aquesta activitat necessitem instal·lar una llibreria específica: executa la següent cel·la

In [ ]:
%pip install openpyxl

In [88]:

import pandas as pd
#TODO


### D) "Taula de Naixements"
Font: https://ibestat.caib.es/ibestat/estadistiques/poblacio/naixements/414cab4f-b402-4cd1-af05-6617443de384 <br/>
URL dades: "https://ibestat.caib.es/ibestat/estadistiques/poblacio/naixements/414cab4f-b402-4cd1-af05-6617443de384" <br/>



In [89]:
import pandas as pd
#TODO


### E) "Nights spent at tourist accommodation establishments by residents/non-residents - monthly data"
Font:  https://ec.europa.eu/eurostat/databrowser/view/tin00171/default/table?lang=en <br/>
Fitxer: "data/tin00171_linear.csv.gz" <br/>
Nota: 
- Els fitxers comprimits amb format .gz podem obrir-se directament com si fossin fitxers de dades amb pandas, i en aquest cas és un fitxer del tipus CSV. És a dir, no cal descomprimir.


In [90]:
import pandas as pd
#TODO

## Activitat 1.2 del homework

[![License: CC BY 4.0](https://img.shields.io/badge/License-CC_BY_4.0-lightgrey.svg)](https://creativecommons.org/licenses/by/4.0/) <br/>
Isaac Lera and Gabriel Moya <br/>
Universitat de les Illes Balears <br/>
isaac.lera@uib.edu, gabriel.moya@uib.edu